go/vt/tabletserver/gorpcqueryservice/sqlquery.go

```diff
@@ -47,32 +91,57 @@ func (sq *SqlQuery) Rollback(ctx context.Context, session *proto.Session, noOutp
 // Execute is exposing tabletserver.SqlQuery.Execute
 func (sq *SqlQuery) Execute(ctx context.Context, query *proto.Query, reply *mproto.QueryResult) (err error) {
 	defer sq.server.HandlePanic(&err)
-	execErr := sq.server.Execute(callinfo.RPCWrapCallInfo(ctx), query, reply)
-	tabletserver.AddTabletErrorToQueryResult(execErr, reply)
+	tErr := sq.server.Execute(callinfo.RPCWrapCallInfo(ctx), query, reply)
+	tabletserver.AddTabletErrorToQueryResult(tErr, reply)
 	if *tabletserver.RPCErrorOnlyInReply {
 		return nil
 	}
-	return execErr
+	return tErr
 }
 
 // StreamExecute is exposing tabletserver.SqlQuery.StreamExecute
 func (sq *SqlQuery) StreamExecute(ctx context.Context, query *proto.Query, sendReply func(reply interface{}) error) (err error) {
 	defer sq.server.HandlePanic(&err)
-	return sq.server.StreamExecute(callinfo.RPCWrapCallInfo(ctx), query, func(reply *mproto.QueryResult) error {
+	tErr := sq.server.StreamExecute(callinfo.RPCWrapCallInfo(ctx), query, func(reply *mproto.QueryResult) error {
 		return sendReply(reply)
 	})
+	if tErr == nil {
+		return nil
+	}
+	if *tabletserver.RPCErrorOnlyInReply {
+		// If there was an app error, send a QueryResult back with it.
+		qr := new(mproto.QueryResult)
+		tabletserver.AddTabletErrorToQueryResult(tErr, qr)
```

Comment:
It will be tricky to roll this out. The client needs to be updated first to handle both cases. However, rdonly clients may take a much longer to fully updated.
I think the StreamExecute call only returns error if it cannot start the streaming. Maybe you can modify the underlying function, and see if the error can be attached to one of existing message (and existing client could handle it).